platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/MerkleStateRoot.java

```diff
@@ -293,652 +131,199 @@ public T copy() {
-            }
-        }
+                        try {
+                            logger.info(STARTUP.getMarker(), "\nMigrating {} (size: {})...", virtualMapToMigrate.getLabel(), virtualMapToMigrate.size());
+                            long migrationStartTime = System.currentTimeMillis();
+                            VirtualMapMigration.extractVirtualMapData(AdHocThreadManager.getStaticThreadManager(), virtualMapToMigrate, handler, THREAD_COUNT);
+                            long migrationTimeMs = System.currentTimeMillis() - migrationStartTime;
+                            logger.info(STARTUP.getMarker(), "Migration complete for {} took {} ms", virtualMapToMigrate.getLabel(), migrationTimeMs);
+                            logger.info(STARTUP.getMarker(), "New Virtual Map size: {}", virtualMap.size());
+                            totalMigrationTimeMs.addAndGet(migrationTimeMs);
+                            totalMigratedObjects.addAndGet(virtualMapToMigrate.size());
+                        } catch (InterruptedException e) { // TODO: revisit exception handling
+                            throw new RuntimeException(e);
+                        }
 
-        @Override
-        public boolean contains(@NonNull final String stateKey) {
-            return stateMetadata.containsKey(stateKey);
-        }
+                        if (VALIDATE_MIGRATION_FF) {
+                            long validationStartTime = System.currentTimeMillis();
+                            logger.info(STARTUP.getMarker(), "Validating the new Virtual Map contains all data from the KV State {}", virtualMapToMigrate.getLabel());
+                            validateKVStateMigrated(virtualMap, virtualMapToMigrate);
+                            long validationTimeMs = System.currentTimeMillis() - validationStartTime;
+                            logger.info(STARTUP.getMarker(), "Validation complete for the KV State {} took {} ms", virtualMapToMigrate.getLabel(), validationTimeMs);
+                            totalValidationTimeMs.addAndGet(validationTimeMs);
+                        }
+                    });
+
+            logger.info(STARTUP.getMarker(), "Migrating Queue states to the one Virtual Map...");
+
+            IntStream.range(0, getNumberOfChildren())
+                    .mapToObj(this::getChild)
+                    .filter(child -> child instanceof QueueNode<?>)
+                    .map(child -> (QueueNode<?>) child)
+                    .forEach(queueNode -> {
+
+                        final StringLeaf originalLabeled = queueNode.getLeft();
+                        final String label = originalLabeled.getLabel();
+                        final var labelPair = decomposeLabel(label);
+                        final var serviceName = labelPair.key();
+                        final var stateKey = labelPair.value();
+                        final FCQueue<ValueLeaf> originalStore = queueNode.getRight();
```

Comment:
FCQueue is probably redundant here. QueueNode has `iterator()` method, which iterates over all elements in the queue state

Reply:
Its the code from here: https://github.com/hashgraph/hedera-services/pull/15369
I think it is how it is for a reason -- to get the codec from `ValueLeaf` so it can be put to the Virtual Map.